### PR TITLE
[Feature] Add test of get_hooks_info()

### PR DESCRIPTION
**Pull request by songyuc**
## Motivation
This PR is linked to https://github.com/open-mmlab/mmengine/issues/599

## Modification
I added the test of `runner.get_hook_info()` in `tests/test_runner/test_runner.py`

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmengine/hooks/hook.py [hunk 1]
```diff
@@ -416,3 +422,24 @@ def is_last_train_iter(self, runner) -> bool:
             bool: Whether current iteration is the last train iteration.
         """
         return runner.iter + 1 == runner.max_iters
+
+    def get_triggered_stages(self):
```
Comment (HAOCHENYE) [inline mmengine/hooks/hook.py:426]:
```suggestion
    def get_triggered_stages(self) -> list:
```

File: mmengine/hooks/hook.py [hunk 2]
```diff
@@ -11,6 +13,10 @@ class Hook:
     """
 
     priority = 'NORMAL'
+    stages = ('before_run', 'before_train_epoch', 'before_train_iter',
```
Comment (HAOCHENYE) [inline mmengine/hooks/hook.py:16]: Missing `after_train`, `after_val` .etc.

File: tests/test_runner/test_runner.py [hunk 3]
```diff
@@ -2306,3 +2306,49 @@ def test_build_runner(self):
...
+                         cfg=None):
+                super().__init__(model=model, work_dir=work_dir)
+
+            # def setup_env(self, env_cfg):
+            #     pass
+
+        cfg.runner_type = 'CustomHooksRunner'
+        runner = RUNNERS.build(cfg)
+        assert isinstance(runner, CustomHooksRunner)
+        assert '(VERY_HIGH   ) RuntimeInfoHook' in runner.get_hook_info()
```
Comment (HAOCHENYE) [inline tests/test_runner/test_runner.py:2354]: target_str = \
"""
xxx
xxx
xxx
"""

We should check the sequence of hooks at least in one stage. For example, `after_train_iter`.

File: tests/test_runner/test_runner.py [hunk 4]
```diff
@@ -2306,3 +2306,49 @@ def test_build_runner(self):
...
+                         log_level='INFO',
+                         visualizer=None,
+                         default_scope=None,
+                         randomness=dict(seed=None),
+                         experiment_name=None,
+                         cfg=None):
+                super().__init__(model=model, work_dir=work_dir)
+
+            # def setup_env(self, env_cfg):
+            #     pass
```
Comment (HAOCHENYE) [inline tests/test_runner/test_runner.py:2349]: Remove unnecessary comments

File: tests/test_runner/test_runner.py [hunk 5]
```diff
@@ -2306,3 +2306,49 @@ def test_build_runner(self):
...
+                         launcher='none',
+                         env_cfg=dict(dist_cfg=dict(backend='nccl')),
+                         log_processor=None,
+                         log_level='INFO',
+                         visualizer=None,
+                         default_scope=None,
+                         randomness=dict(seed=None),
+                         experiment_name=None,
+                         cfg=None):
+                super().__init__(model=model, work_dir=work_dir)
```
Comment (HAOCHENYE) [inline tests/test_runner/test_runner.py:2346]: Why do we need to define a custom runner?

**Review by HAOCHENYE — APPROVED** [review]

File: .gitignore [hunk 6]
```diff
@@ -121,3 +121,4 @@ work_dirs/
 *.pth
 *.py~
 *.sh~
```
Comment (HAOCHENYE) [inline .gitignore:123]: clean the unnecessary changes

File: mmengine/runner/runner.py [hunk 7]
```diff
@@ -1577,6 +1580,29 @@ def build_log_processor(
 
         return log_processor  # type: ignore
 
+    def get_hook_info(self) -> str:
```
Comment (HAOCHENYE) [inline mmengine/runner/runner.py:1583]:
```suggestion
    def get_hooks_info(self) -> str:
```

File: tests/test_runner/test_runner.py [hunk 8]
```diff
@@ -2306,3 +2306,18 @@ def test_build_runner(self):
...
+        cfg.runner_type = 'Runner'
+        runner = RUNNERS.build(cfg)
+        assert isinstance(runner, Runner)
+        target_str = ('after_train_iter:\n'
+                      '(VERY_HIGH   ) RuntimeInfoHook                    \n'
+                      '(NORMAL      ) IterTimerHook                      \n'
+                      '(BELOW_NORMAL) LoggerHook                         \n'
+                      '(LOW         ) ParamSchedulerHook                 \n'
+                      '(VERY_LOW    ) CheckpointHook                     \n')
+        assert target_str in runner.get_hook_info()
```
Comment (HAOCHENYE) [inline tests/test_runner/test_runner.py:2323]: suggest using `self.assertIn` and `self.assertIsinstance`

**Review by HAOCHENYE — DISMISSED** [review]

File: mmengine/hooks/hook.py [hunk 9]
```diff
@@ -416,3 +425,24 @@ def is_last_train_iter(self, runner) -> bool:
...
+                trigger_stages.add(stage)
+
+        # some methods will be triggered in multi stages
+        # use this dict to map method to stages.
+        method_stages_map = {
+            '_before_epoch': ['before_train_epoch', 'before_val_epoch'],
+            '_after_epoch': ['after_train_epoch', 'after_val_epoch'],
+            '_before_iter': ['before_train_iter', 'before_val_iter'],
+            '_after_iter': ['after_train_iter', 'after_val_iter'],
+        }
```
Comment (HAOCHENYE) [inline mmengine/hooks/hook.py:442]: Missing `test`

File: mmengine/hooks/hook.py [hunk 10]
```diff
@@ -416,3 +425,24 @@ def is_last_train_iter(self, runner) -> bool:
...
+
+        for method, map_stages in method_stages_map.items():
+            if is_method_overridden(method, Hook, self):
+                trigger_stages.update(map_stages)
+
+        return [stage for stage in Hook.stages if stage in trigger_stages]
```
Comment (HAOCHENYE) [inline mmengine/hooks/hook.py:448]: Why don't we return `list(trigger_stages)` directly?

File: mmengine/runner/runner.py [hunk 11]
```diff
@@ -413,6 +413,9 @@ def __init__(
         self._hooks: List[Hook] = []
         # register hooks to `self._hooks`
         self.register_hooks(default_hooks, custom_hooks)
+        # log hooks information
+        self.logger.info('Hooks will be executed in the following order:\n%s',
+                         self.get_hooks_info())
```
Comment (HAOCHENYE) [inline mmengine/runner/runner.py:418]:
```suggestion
        self.logger.info(f'Hooks will be executed in the following order:\n{self.get_hooks_info()}')
```

**Comment by zhouzaida** [comment]
Hi @songyuc , thanks for your contribution. One thing to note is that MMEngine support adding additional custom hook methods so this logic for printing hook information is a little different from MMCV.

https://github.com/open-mmlab/mmengine/blob/79c8d735a0615537d1d618c0dd6453d5ee146402/mmengine/runner/runner.py#L1735-L1749

**Comment by HAOCHENYE** [comment]
> Hi @songyuc , thanks for your contribution. One thing to note is that MMEngine support adding additional custom hook methods so this logic for printing hook information is a little different from MMCV.
> 
> https://github.com/open-mmlab/mmengine/blob/79c8d735a0615537d1d618c0dd6453d5ee146402/mmengine/runner/runner.py#L1735-L1749

Is it reasonable to call `get_hooks_info` in `RuntimeInfoHook.before_run`?

**Review by zhouzaida — APPROVED** [review]

**Review by RangiLyu — APPROVED** [review]
